plugins/woocommerce-admin/client/products/fields/attribute-field/attribute-field.tsx

```diff
@@ -82,29 +94,51 @@ export const AttributeField: React.FC< AttributeFieldProps > = ( {
 	);
 
 	useEffect( () => {
-		if ( ! value || hydrationComplete ) {
+		// Temporarily always doing hydration, since otherwise new attributes
+		// get removed from Options and Attributes when the other list is then
+		// modified
+		//
+		// This is because the hydration is out of date -- the logic currently
+		// assumes modifications are only made from within the component
+		//
+		// I think we'll need to move the hydration out of the individual component
+		// instance. To where, I do not yet know... maybe in the form context
+		// somewhere so that a single hydration source can be shared between multiple
+		// instances? Something like a simple key-value store in the form context
+		// would be handy.
+		if ( ! value /*|| hydrationComplete*/ ) {
 			return;
 		}
 
 		const [ customAttributes, globalAttributes ]: ProductAttribute[][] =
 			sift( value, ( attr: ProductAttribute ) => attr.id === 0 );
 
 		Promise.all(
-			globalAttributes.map( ( attr ) => fetchTerms( attr.id ) )
+			globalAttributes.map( ( attr ) => {
+				return fetchTerms( attr.id );
+			} )
 		).then( ( allResults ) => {
 			setHydratedAttributes( [
 				...globalAttributes.map( ( attr, index ) => {
+					const fetchedTerms = allResults[ index ];
+
 					const newAttr = {
 						...attr,
-						terms: allResults[ index ],
-						options: undefined,
+						// I'm not sure this is quite right for handling unpersisted terms,
+						// but this gets things kinda working for now
+						terms:
+							fetchedTerms.length > 0 ? fetchedTerms : undefined,
+						options:
+							fetchedTerms.length === 0
+								? attr.options
+								: undefined,
```

Comment:
This should be revisited... depending on how hydration is handled, it might not be needed.